lib/actions-oclif/env/add.ts

```diff
@@ -171,31 +190,60 @@ async function setServiceVars(
 	options: FlagsDef,
 ) {
 	if (options.application) {
-		const serviceId = await getServiceIdForApp(
-			sdk,
-			options.application,
-			options.service!,
-		);
-		await sdk.models.service.var.set(serviceId, params.name, params.value!);
-	} else {
+		for (const app of options.application.split(',')) {
+			for (const service of options.service!.split(',')) {
+				try {
+					const serviceId = await getServiceIdForApp(sdk, app, service);
+					await sdk.models.service.var.set(
+						serviceId,
+						params.name,
+						params.value!,
+					);
+				} catch (err) {
+					console.error(`${err.message}, application: ${app}`);
+					process.exitCode = 1;
+				}
+			}
+		}
+	} else if (options.device) {
 		const { getDeviceAndAppFromUUID } = await import('../../utils/cloud');
-		const [device, app] = await getDeviceAndAppFromUUID(
-			sdk,
-			options.device!,
-			['id'],
-			['app_name'],
-		);
-		const serviceId = await getServiceIdForApp(
-			sdk,
-			app.app_name,
-			options.service!,
-		);
-		await sdk.models.device.serviceVar.set(
-			device.id,
-			serviceId,
-			params.name,
-			params.value!,
-		);
+		for (const uuid of options.device.split(',')) {
+			let device;
+			let app;
+			try {
+				[device, app] = await getDeviceAndAppFromUUID(
+					sdk,
+					uuid,
+					['id'],
+					['app_name'],
+				);
+			} catch (err) {
+				console.error(`${err.message}, device: ${uuid}`);
+				process.exitCode = 1;
+			}
+			for (const service of options.service!.split(',')) {
+				try {
+					if (app) {
+						const serviceId = await getServiceIdForApp(
+							sdk,
+							app.app_name,
+							service,
+						);
+						if (device) {
+							await sdk.models.device.serviceVar.set(
+								device.id,
+								serviceId,
+								params.name,
+								params.value!,
+							);
+						}
+					}
```

Comment:
Adding the `return` in the catch will avoid the need for the `if (app)` below and also allow the process to exit straight away
```suggestion
			} catch (err) {
				console.error(`${err.message}, device: ${uuid}`);
				process.exitCode = 1;
				return;
			}
			for (const service of options.service!.split(',')) {
				try {
					const serviceId = await getServiceIdForApp(
						sdk,
						app.app_name,
						service,
					);
					if (device) {
						await sdk.models.device.serviceVar.set(
							device.id,
							serviceId,
							params.name,
							params.value!,
						);
					}
```

Reply:
I don't think we would want to return at this point. In a previous PR for these bulk commands, @pdcastro and I discussed what happens if there is a problem with one value in a list of devices, services, or apps. If we return due to an error with a value in the middle of the list, then the values before the error value will have been changed and the values after the error would not have been looked at. We decided it was better for the user to continue changing the remaining values and report the values that caused an error.

Reply:
[Original discussion](https://github.com/balena-io/balena-cli/pull/1939#discussion_r453690066)

Reply:
Oh right we're in a loop here, what about using a `continue` to skip just the current uuid that failed:
```suggestion
			} catch (err) {
				console.error(`${err.message}, device: ${uuid}`);
				process.exitCode = 1;
				continue;
			}
			for (const service of options.service!.split(',')) {
				try {
					const serviceId = await getServiceIdForApp(
						sdk,
						app.app_name,
						service,
					);
					if (device) {
						await sdk.models.device.serviceVar.set(
							device.id,
							serviceId,
							params.name,
							params.value!,
						);
					}
```

Reply:
That would be fine with me, though the behavior is the same either way(based on previous testing). I think the `continue` would explicitly show the behavior when reading the code.

Reply:
I believe the `continue` statement as suggested by Page is best practice ("almost a bug" not to use it), because there is no purpose in even entering the inner loop if the `device` and `app` variables are not successfully set being set by the outer loop. To me it feels "disrespectful of the CPU and the science of algorithm analysis" 😁  to enter a loop just to _repeatedly_ run the same tests that skip all code execution, as opposed to not entering the loop in the first place if it is practical to avoid it. I know, you would rightly argue that the "performance" difference is negligible in micro or nanoseconds, but it's a matter of principle and style! 😎 

Btw, Page's suggested snippet still leaves behind the `if (device)` test in the inner loop, but it looks redundant and should also be removed when the `continue` statement is added. As a bonus, this change would also improve legibility a little bit, and reduce some 4 lines of code.

Reply:
I was thinking too much about the inner loop. This does make sense and for cleaner code.

Reply:
Sorry for my mouthful / storm in a teacup. 😁   It looks better now with the `continue` statement. 👍
Thank you!

Reply:
No Worries! I was thinking about it wrong. Thanks for the correction! I completely agree after looking closer.